### PR TITLE
Biometric authentication

### LINKED ISSUE #61: Biometric authentication

### PR DESCRIPTION
**Pull request by louischan-oursky**
fix #61

### CODE REVIEW

File: sdk/src/main/java/com/oursky/authgear/AuthgearCore.kt [hunk 1]
```diff
@@ -558,6 +568,208 @@ internal class AuthgearCore(
...
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    suspend fun enableBiometric(
+        options: BiometricOptions
+    ) {
+        requireIsInitialized()
+
+        val accessToken: String = this.accessToken
+            ?: throw IllegalStateException("enableBiometric required authenticated user.")
```
Comment (carmenlau) [inline sdk/src/main/java/com/oursky/authgear/AuthgearCore.kt:616]: Is there a chance that when user call enableBiometric the access token expired already? It is handled?

File: javasample/src/main/java/com/oursky/authgeartest/MainActivity.java [hunk 2]
```diff
@@ -112,13 +123,19 @@ private void updateButtonDisabledState(MainViewModel viewModel) {
         UserInfo userInfo = viewModel.userInfo().getValue();
         boolean isLoading = viewModel.isLoading().getValue();
         boolean isConfigured = viewModel.isConfigured().getValue();
-        boolean isLoggedIn = viewModel.isLoggedIn().getValue();
         boolean isAnonymous = userInfo != null && userInfo.isAnonymous();
+        boolean isBiometricSupported = viewModel.isBiometricSupported().getValue();
+        boolean isBiometricEnabled = viewModel.isBiometricEnabled().getValue();
+        boolean isLoggedIn = userInfo != null;
```
Comment (carmenlau) [inline javasample/src/main/java/com/oursky/authgeartest/MainActivity.java:129]: The buttons isLoggedIn state is incorrect. When I 

1. call authorize login success
2. kill the app
3. reopen the app
4. call configure

I should logged in, but the UI show me I am not.

**Review by carmenlau — COMMENTED** [review]
The app crash when I try enable biometric in the example. My device is Nokia 7.2 with Android 10.

Here is the exception
```
2021-03-25 11:27:24.802 26768-26821/com.oursky.authgeartest E/AndroidRuntime: FATAL EXCEPTION: DefaultDispatcher-worker-2
    Process: com.oursky.authgeartest, PID: 26768
    java.lang.NoSuchMethodError: No virtual method setUserAuthenticationParameters(II)Landroid/security/keystore/KeyGenParameterSpec$Builder; in class Landroid/security/keystore/KeyGenParameterSpec$Builder; or its super classes (declaration of 'android.security.keystore.KeyGenParameterSpec$Builder' appears in /system/framework/framework.jar!classes2.dex)
        at com.oursky.authgear.BiometricKt.makeGenerateKeyPairSpec(Biometric.kt:26)
        at com.oursky.authgear.AuthgearCore.enableBiometric(AuthgearCore.kt:630)
        at com.oursky.authgear.Authgear$enableBiometric$1.invokeSuspend(Authgear.kt:487)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
2021-03-25 11:27:24.815 846-846/? E/lowmemorykiller: Error writing /proc/26768/oom_score_adj; errno=22
```

File: javasample/src/main/java/com/oursky/authgeartest/MainActivity.java (hunk 2 above)
Comment (louischan-oursky) [reply javasample/src/main/java/com/oursky/authgeartest/MainActivity.java:129]: The iOS example app has the same behavior. Do you want me to correct all SDK examples?

File: sdk/src/main/java/com/oursky/authgear/AuthgearCore.kt (hunk 1 above)
Comment (louischan-oursky) [reply sdk/src/main/java/com/oursky/authgear/AuthgearCore.kt:616]: Not handled. It is actually another problem. Both the iOS SDK and Android has this issue. The handling of refreshing access token is not automatic. That's why I proposed to improve the SDK before soft-launch.

**Comment by louischan-oursky** [comment]
Added API level checking and switched to use `sessionState`.

**Review by carmenlau — COMMENTED** [review]
The API version checking works great!

But I found that the error is confusing when I have not enabled biometric in server. iOS can show me `BiometricDisallowed`.

device-2021-03-26-121603

**Comment by louischan-oursky** [comment]
The io error revealed a big problem in the SDK..... Anyway I fixed that.

**Review by carmenlau — COMMENTED** [review]
Thanks for adding the error handling!! Only some question about `makeError` function. The SDK overall looks good.

But during my testing, I found some problem in server `biometric` branch. When I try promote a user, I failed to visit the promote endpoint with too many redirect error. See if you encounter the same problem, may need address it in server.

File: sdk/src/main/java/com/oursky/authgear/data/HttpClient.kt [hunk 3]
```diff
@@ -1,93 +1,84 @@
...
-        fun postForm(url: URL, body: MutableMap<String, String>) {
-            val conn = url.openConnection() as HttpURLConnection
-            try {
-                conn.doOutput = true
-                conn.requestMethod = "POST"
-                conn.setRequestProperty("content-type", "application/x-www-form-urlencoded")
-                conn.outputStream.use {
-                    it.write(body.toFormData().toByteArray(StandardCharsets.UTF_8))
+        fun makeError(jsonObject: JSONObject): Exception? {
+            if (jsonObject.has("name") && jsonObject.has("reason") && jsonObject.has("message")) {
```
[the remark this answers is not in the document]
Comment (louischan-oursky) [reply sdk/src/main/java/com/oursky/authgear/data/HttpClient.kt:48]: I copied the logic from the JS SDK. The shape of api error should be correct.

[the remark this answers is not in the document]
Comment (carmenlau) [reply sdk/src/main/java/com/oursky/authgear/data/HttpClient.kt:48]: Should check key `error`, API error payload should be inside response object with key `error`.

**Comment by louischan-oursky** [comment]
Fixed the error decoding.

**Pull request merged**